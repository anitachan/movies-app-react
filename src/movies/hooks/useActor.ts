import { useEffect, useState } from 'react';
import { Actor } from '../models/actor.model';
import { ActorCast } from '../models/actorCast.model';
import { getActor, getActorCredits } from '../services/tmdbApi';

export const useActor = (id: string) => {
  const [actor, setActor] = useState<Actor>();
  const [credits, setCredits] = useState<ActorCast[]>();
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchActor = async () => {
      try {
        const [resultActor, resultActorCast] = await Promise.all([
          getActor(id),
          getActorCredits(id),
        ]);
        setActor(resultActor);
        setCredits(resultActorCast);
      } catch (error: any) {
        setError(error.message || 'Error loading actor');
      } finally {
        setLoading(false);
      }
    };

    fetchActor();
  }, [id]);

  return { actor, credits, loading, error };
};
