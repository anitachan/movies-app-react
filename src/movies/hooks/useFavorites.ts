import { useEffect, useState } from 'react';
import { MovieDetail } from '../models/movieDetail';

export const useFavorites = () => {
  const [favorites, setFavorites] = useState<MovieDetail[]>([]);
  useEffect(() => {
    try {
      const getLocalFavorites = localStorage.getItem('favorites');
      const storedFavorites: MovieDetail[] = getLocalFavorites ? JSON.parse(getLocalFavorites) : [];
      setFavorites(storedFavorites);
    } catch (error) {
      setFavorites([]);
    }
  }, []);

  return { favorites };
};
