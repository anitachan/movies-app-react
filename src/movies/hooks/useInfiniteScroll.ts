import { useEffect, useRef } from "react";

export const useInfiniteScroll = (callback: () => void) => {
  const observerRef = useRef<IntersectionObserver | null>(null);
  const loaderRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (!loaderRef.current) return;

    observerRef.current = new IntersectionObserver(([entry]) => {
      if (entry.isIntersecting) {
        callback();
      }
    }, { threshold: 0.5 });

    observerRef.current.observe(loaderRef.current);

    return () => observerRef.current?.disconnect();
  }, [callback]);

  return loaderRef;
};
