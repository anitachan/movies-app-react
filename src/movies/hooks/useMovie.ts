import { useEffect, useState } from "react";
import { Cast } from "../models/cast.model";
import { MovieDetail } from "../models/movieDetail";
import { getCast, getMovie } from "../services/tmdbApi";

export const useMovie = (id: string) => {
    const [movie, setMovie] = useState<MovieDetail>();
    const [cast, setCast] = useState<Cast[]>();
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);

    useEffect(() => {
        const fetchMovie = async () => {
            try {
                const [movieResult, castResult] = await Promise.all([getMovie(id), getCast(id)]);
                setMovie(movieResult);
                setCast(castResult);
            } catch (error: any) {
                setError(error.message || 'Loading movie error')
            } finally {
                setLoading(false);
            }
        }

        fetchMovie();
    }, [id])

    return { movie, cast, loading, error };
};