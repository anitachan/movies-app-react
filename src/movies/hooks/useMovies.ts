import { useEffect, useRef, useState } from "react";
import type { Movie } from "../models/movies.model";
import { getMovies } from "../services/tmdbApi";

export const useMovies = () => {
    const [movies, setMovies] = useState<Movie[]>([]);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState<string | null>(null);
    const [page, setPage] = useState(1);
    const hasFetchedInitial = useRef(false)

    useEffect(() => {
        const fetchMovies = async () => {
            if (loading) return;

            setLoading(true);
            try {
                const result = await getMovies(page);
                setMovies((prev) => [...prev, ...result]);
            } catch {
                setError("Loading movies error");
            } finally {
                setLoading(false);
            }
        };

        if (page === 1 && hasFetchedInitial.current) return;
        hasFetchedInitial.current = true;

        fetchMovies();
    }, [page, loading]);

    const loadMore = () => {
        if (loading) return;
        setPage((prev) => prev + 1);
    };

    return { movies, loading, error, loadMore };
};
