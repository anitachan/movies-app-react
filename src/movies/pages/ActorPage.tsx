import { useParams } from 'react-router-dom';
import { Box, Typography, Chip, Link } from '@mui/material';
import { useActor } from '../hooks/useActor';
import noUser from '../../assets/no-user.png';
import { CreditsCard } from '../components/CreditsCard';

export const ActorPage = () => {
	const { id } = useParams();
	const { actor, credits } = useActor(id!);
	const url = process.env.REACT_APP_IMAGES_URL!;
	const photoSrc = actor?.profile_path ? `${url}${actor.profile_path}` : noUser;

	return (
		<Box
			component="main"
			sx={{
				width: '100%',
				boxSizing: 'border-box',
				p: 2,
			}}
		>
			<Box
				sx={{
					display: 'grid',
					gap: 2,
					gridTemplateColumns: 'repeat(12, 1fr)',
					gridAutoRows: 'auto',
				}}
			>
				<Box
					component="img"
					src={photoSrc}
					alt={actor?.name}
					loading="lazy"
					onError={(e) => {
						e.currentTarget.onerror = null;
						e.currentTarget.src = noUser;
					}}
					sx={{
						gridColumn: 'span 4',
						width: '100%',
						height: 'auto',
						objectFit: 'cover',
						borderRadius: 2,
						display: 'block',
					}}
				/>
				<Box
					sx={{
						gridColumn: 'span 8',
						minWidth: 0,
					}}
				>
					<Typography variant="h4" gutterBottom>
						{actor?.name}
					</Typography>
					<Typography variant="body2" color="text.secondary" gutterBottom>
						{actor?.known_for_department}
					</Typography>
					<Typography variant="body2" gutterBottom>
						<strong>Born:</strong> {actor?.birthday ? new Date(actor.birthday).toLocaleDateString() : 'â€”'} in{' '}
						{actor?.place_of_birth}
					</Typography>
					{actor?.deathday && (
						<Typography variant="body2" gutterBottom>
							<strong>Died:</strong> {new Date(actor.deathday).toLocaleDateString()}
						</Typography>
					)}
					{actor?.homepage && (
						<Typography variant="body2" gutterBottom>
							<strong>Homepage:</strong>{' '}
							<Link href={actor.homepage} target="_blank" rel="noopener">
								{actor.homepage}
							</Link>
						</Typography>
					)}

					<Box sx={{ mt: 2, display: 'flex', flexWrap: 'wrap', gap: 1 }}>
						{actor?.also_known_as.map((aka) => (
							<Chip key={aka} label={aka} size="small" />
						))}
					</Box>

					<Box sx={{ mt: 4 }}>
						<Typography variant="h6" gutterBottom>
							Biography
						</Typography>
						<Typography variant="body1" sx={{ whiteSpace: 'pre-line' }}>
							{actor?.biography}
						</Typography>
					</Box>
				</Box>
				<Box sx={{ gridColumn: '1 / -1', mt: 4 }}>
					<Typography variant="h6" gutterBottom>
						Credits
					</Typography>
					<Box
						sx={{
							display: 'flex',
							gap: 2,
							overflowX: 'auto',
							overflowY: 'hidden',
							py: 1,
							'&::-webkit-scrollbar': { height: 6 },
							'&::-webkit-scrollbar-thumb': {
								backgroundColor: 'rgba(0,0,0,0.3)',
								borderRadius: 3,
							},
						}}
					>
						{credits?.map((credit) => (
							<Box key={credit.credit_id} sx={{ flex: '0 0 auto' }}>
								<CreditsCard credit={credit!} />
							</Box>
						))}
					</Box>
				</Box>
			</Box>
		</Box>
	);
};
