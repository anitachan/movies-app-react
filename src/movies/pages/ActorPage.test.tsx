import { fireEvent, render, screen } from '@testing-library/react';
import { createList, fixtureOf } from 'ts-mock-autofixture-kit';
import noUser from '../../assets/no-user.png';
import { useActor } from '../hooks/useActor';
import { Actor } from '../models/actor.model';
import { ActorCast } from '../models/actorCast.model';
import { ActorPage } from './ActorPage';

process.env.REACT_APP_IMAGES_URL ||= 'https://example.com/';

jest.mock('../hooks/useActor', () => ({
  __esModule: true,
  useActor: jest.fn(),
}));

jest.mock('../components/CreditsCard', () => ({
  __esModule: true,
  CreditsCard: ({ credit }: { credit: any }) => (
    <div data-testid="mock-credits-card">{credit?.character}</div>
  ),
}));

jest.mock(
  'react-router-dom',
  () => ({
    useParams: () => ({ id: 'some-actor-id' }),
  }),
  { virtual: true }
);

describe('ActorPage', () => {
  const useActorMock = useActor as jest.MockedFunction<typeof useActor>;
  const mockActor = fixtureOf<Actor>({ seed: 42 }).create();
  const mockCredits = createList<ActorCast>(3, { seed: 100 });
  beforeEach(async () => {
    useActorMock.mockReturnValue({
      actor: mockActor,
      credits: mockCredits,
      error: null,
      loading: false,
    });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render actor and show credits', () => {
    const { container } = render(<ActorPage />);

    expect(container).toMatchSnapshot();

    const cards = screen.getAllByTestId('mock-credits-card');
    expect(cards).toHaveLength(3);
    expect(cards[0].textContent).toBe(mockCredits[0].character);
    expect(cards[1].textContent).toBe(mockCredits[1].character);
    expect(cards[2].textContent).toBe(mockCredits[2].character);
  });

  test('should set noUser when poster is empty', () => {
    const mockActorPoster = fixtureOf<Actor>()
      .with('profile_path', () => '')
      .with('name', () => 'No Poster Actor')
      .create();

    useActorMock.mockReturnValue({
      actor: mockActorPoster,
      credits: mockCredits,
      error: null,
      loading: false,
    });

    render(<ActorPage />);
    const image = screen.getByAltText('No Poster Actor') as HTMLImageElement;

    expect(image).toBeInTheDocument();
    expect(image.src).toContain(noUser);
  });

  test('should set no user when image has an error ', () => {
    render(<ActorPage />);
    const img = screen.getByAltText(mockActor.name) as HTMLImageElement;

    expect(img.src).toContain(mockActor.profile_path);

    fireEvent.error(img);
    expect(img.src).toContain('no-user.png');
  });

  test('should show a dash when birthday date is empty', () => {
    const actorNoBirthday = fixtureOf<Actor>()
      .with('birthday', () => undefined)
      .with('name', () => 'no Birthday')
      .with('place_of_birth', () => 'city')
      .create();

    useActorMock.mockReturnValue({
      actor: actorNoBirthday,
      credits: mockCredits,
      loading: false,
      error: null,
    });

    render(<ActorPage />);

    const bornStrong = screen.getByText('Born:');
    const bornParagraph = bornStrong.closest('p');
    expect(bornParagraph).toBeInTheDocument();

    expect(bornParagraph).toHaveTextContent(`Born: â€” in city`);
  });
});
