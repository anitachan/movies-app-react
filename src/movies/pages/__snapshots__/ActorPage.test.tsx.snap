// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActorPage should render actor and show credits 1`] = `
<div>
  <main
    class="MuiBox-root css-vuvul"
  >
    <div
      class="MuiBox-root css-3jp51e"
    >
      <img
        alt="3CpplugYH0Te"
        class="MuiBox-root css-e19die"
        loading="lazy"
        src="https://image.tmdb.org/t/p/originalaGTqxm8FGpgY"
      />
      <div
        class="MuiBox-root css-huqchk"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
        >
          3CpplugYH0Te
        </h4>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-1bub9p-MuiTypography-root"
        >
          Hi1CDBoab94W
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ytim1c-MuiTypography-root"
        >
          <strong>
            Born:
          </strong>
           
          Invalid Date
           in
           
          Cule9Hf9ARlH
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ytim1c-MuiTypography-root"
        >
          <strong>
            Died:
          </strong>
           
          Invalid Date
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-ytim1c-MuiTypography-root"
        >
          <strong>
            Homepage:
          </strong>
           
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1orsvah-MuiTypography-root-MuiLink-root"
            href="NxpFGg8vVqP4"
            rel="noopener"
            target="_blank"
          >
            NxpFGg8vVqP4
          </a>
        </p>
        <div
          class="MuiBox-root css-2w3808"
        >
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-1lflu6f-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
            >
              B5C8tX1UIZGt
            </span>
          </div>
          <div
            class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorDefault MuiChip-filledDefault css-1lflu6f-MuiChip-root"
          >
            <span
              class="MuiChip-label MuiChip-labelSmall css-eccknh-MuiChip-label"
            >
              vtBJOdECaRyB
            </span>
          </div>
        </div>
        <div
          class="MuiBox-root css-h5fkc8"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
          >
            Biography
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-c04sfq-MuiTypography-root"
          >
            6hAszpwS6bDY
          </p>
        </div>
      </div>
      <div
        class="MuiBox-root css-1kk4q3i"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
        >
          Credits
        </h6>
        <div
          class="MuiBox-root css-j510m5"
        >
          <div
            class="MuiBox-root css-1tu59u4"
          >
            <div
              data-testid="mock-credits-card"
            >
              fCv8hVvQUP3c
            </div>
          </div>
          <div
            class="MuiBox-root css-1tu59u4"
          >
            <div
              data-testid="mock-credits-card"
            >
              5EyEWWSUkwE8
            </div>
          </div>
          <div
            class="MuiBox-root css-1tu59u4"
          >
            <div
              data-testid="mock-credits-card"
            >
              VGCbCWgYKde4
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
